src/middleware.ts

```diff
@@ -61,6 +61,72 @@ function binaryToJsonMessageStream<T extends JSONRPCMessage>(
   });
 }
 
+/**
+ * This function is a factory to create a TransformStream that will
+ * transform a `Uint8Array` stream to a stream containing the JSON header
+ * message and the rest of the data in `Uint8Array` format.
+ * The header message will be validated with the provided messageParser, this
+ * also infers the type of the stream output.
+ * @param messageParser - Validates the JSONRPC messages, so you can select for a
+ *  specific type of message
+ * @param bufferByteLimit - sets the number of bytes buffered before throwing an
+ *  error. This is used to avoid infinitely buffering the input.
+ */
+function binaryToJsonHeaderMessageStream<T extends JSONRPCMessage>(
+  messageParser: (message: unknown) => T,
+  bufferByteLimit: number = 1024 * 1024,
+): TransformStream<Uint8Array, T | Uint8Array> {
+  const parser = new JSONParser({
+    separator: '',
+    paths: ['$'],
+  });
+  let bytesWritten: number = 0;
+  let accumulator = Buffer.alloc(0);
+  let rawStream = false;
+
+  return new TransformStream<Uint8Array, T | Uint8Array>({
+    flush: async () => {
+      // Avoid potential race conditions by allowing parser to end first
+      const waitP = utils.promise();
+      parser.onEnd = () => waitP.resolveP();
+      parser.end();
+      await waitP.p;
```

Comment:
The parser should be cleaned up after the first message. but also keep in mind that the stream can end before the first message so you need to clean up in that case as well.